ddtrace/opentelemetry/span_test.go

```diff
@@ -22,63 +24,248 @@ import (
+	var p string
+	select {
+	case <-ctx.Done():
+		return "", fmt.Errorf("Timed out waiting for traces")
+	case p = <-payloads:
+		break
+	}
+	return p, nil
+}
+
+func TestSpanSetName(t *testing.T) {
+	assert := assert.New(t)
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
 
-	otel.SetTracerProvider(NewTracerProvider(
-		tracer.WithEnv(testData.env),
-		tracer.WithHTTPClient(c),
-		tracer.WithService(testData.srv)))
+	_, payloads, cleanup := mockTracerProvider(t)
 	tr := otel.Tracer("")
+	defer cleanup()
+
+	_, sp := tr.Start(ctx, "OldName")
+	sp.SetName("NewName")
+	sp.End()
+
+	tracer.Flush()
+	p, err := waitForPayload(ctx, payloads)
+	if err != nil {
+		t.Log(err)
+		t.Fail()
+	}
+	assert.Contains(p, "NewName")
+}
+
+func TestSpanEnd(t *testing.T) {
 	assert := assert.New(t)
+	testData := []struct {
+		trueName        string
+		falseName       string
+		trueError       codes.Code
+		trueErrorMsg    string
+		falseError      codes.Code
+		falseErrorMsg   string
+		trueAttributes  map[string]string
+		falseAttributes map[string]string
+	}{
+		{
+			trueName:        "trueName",
+			falseName:       "invalidName",
+			trueError:       codes.Error,
+			trueErrorMsg:    "error_description",
+			falseError:      codes.Ok,
+			falseErrorMsg:   "ok_description",
+			trueAttributes:  map[string]string{"trueKey": "trueVal"},
+			falseAttributes: map[string]string{"trueKey": "fakeVal", "invalidKey": "invalidVal"},
+		},
+	}
+	_, payloads, cleanup := mockTracerProvider(t)
+	tr := otel.Tracer("")
+	defer cleanup()
+
+	for _, test := range testData {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		_, sp := tr.Start(context.Background(), test.trueName)
+		sp.SetStatus(codes.Error, test.trueErrorMsg)
+		for k, v := range test.trueAttributes {
+			sp.SetAttributes(attribute.String(k, v))
+		}
+		assert.True(sp.IsRecording())
+		sp.End()
+		assert.False(sp.IsRecording())
+		// following operations should not be able to modify the Span
+		sp.SetName(test.trueName)
+		sp.SetStatus(test.falseError, test.falseErrorMsg)
+		for k, v := range test.trueAttributes {
+			sp.SetAttributes(attribute.String(k, v))
+			sp.SetAttributes(attribute.String(k, v))
+		}
 
-	_, sp := tr.Start(context.Background(), testData.oldOp)
-	for _, tag := range testData.tags {
+		tracer.Flush()
+		payload, err := waitForPayload(ctx, payloads)
+		if err != nil {
+			t.Log(err)
+			t.Fail()
+		}
+		assert.Contains(payload, test.trueErrorMsg)
+		assert.NotContains(payload, test.falseErrorMsg)
+		assert.Contains(payload, test.trueName)
+		assert.NotContains(payload, test.falseName)
+		for k, v := range test.trueAttributes {
+			assert.Contains(payload, fmt.Sprintf("\"%s\":\"%s\"", k, v))
+		}
+		for k, v := range test.falseAttributes {
+			assert.NotContains(payload, fmt.Sprintf("\"%s\":\"%s\"", k, v))
+		}
+		cancel()
+	}
+}
+
+func TestSpanSetStatus(t *testing.T) {
```

Comment:
I suggest a comment at the beginning of this test explaining that this is verifying the spec's behavior around status, namely that the description is set iff the status is an error, and that if SetStatus is called multiple times, the correct precedence is used, with a link to the spec https://opentelemetry.io/docs/reference/specification/trace/api/#set-status